73_ArithmeticBench/main.cpp

```diff
@@ -0,0 +1,949 @@
+			return logFail("Could not create Window & Surface or initialize the Surface!");
+
+		auto pool = m_device->createCommandPool(graphicsQueue->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+
+		for (auto i = 0u; i < MaxFramesInFlight; i++)
+		{
+			if (!pool)
+				return logFail("Couldn't create Command Pool!");
+			if (!pool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, { m_cmdBufs.data() + i, 1 }))
+				return logFail("Couldn't create Command Buffer!");
+		}
+
+		m_winMgr->setWindowSize(m_window.get(), WIN_W, WIN_H);
+		m_surface->recreateSwapchain();
+
+		transferDownQueue = getTransferDownQueue();
+		computeQueue = getComputeQueue();
+
+		// TODO: get the element count from argv
+		const uint32_t elementCount = Output<>::ScanElementCount;
+		// populate our random data buffer on the CPU and create a GPU copy
+		inputData = new uint32_t[elementCount];
+		smart_refctd_ptr<IGPUBuffer> gpuinputDataBuffer;
+		{
+			std::mt19937 randGenerator(0xdeadbeefu);
+			for (uint32_t i = 0u; i < elementCount; i++)
+				inputData[i] = randGenerator(); // TODO: change to using xoroshiro, then we can skip having the input buffer at all
+
+			IGPUBuffer::SCreationParams inputDataBufferCreationParams = {};
+			inputDataBufferCreationParams.size = sizeof(Output<>::data[0]) * elementCount;
+			inputDataBufferCreationParams.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+			m_utils->createFilledDeviceLocalBufferOnDedMem(
+				SIntendedSubmitInfo{.queue=getTransferUpQueue()},
+				std::move(inputDataBufferCreationParams),
+				inputData
+			).move_into(gpuinputDataBuffer);
+		}
+
+		// create 8 buffers for 8 operations
+		for (auto i=0u; i<OutputBufferCount; i++)
+		{
+			IGPUBuffer::SCreationParams params = {};
+			params.size = sizeof(uint32_t) + gpuinputDataBuffer->getSize();
+			params.usage = bitflag(IGPUBuffer::EUF_STORAGE_BUFFER_BIT) | IGPUBuffer::EUF_TRANSFER_SRC_BIT;
+
+			outputBuffers[i] = m_device->createBuffer(std::move(params));
+			auto mreq = outputBuffers[i]->getMemoryReqs();
+			mreq.memoryTypeBits &= m_physicalDevice->getDeviceLocalMemoryTypeBits();
+			assert(mreq.memoryTypeBits);
+
+			auto bufferMem = m_device->allocate(mreq, outputBuffers[i].get());
+			assert(bufferMem.isValid());
+		}
+
+		// create dummy image
+		dummyImg = m_device->createImage({
+				{
+					.type = IGPUImage::ET_2D,
+					.samples = asset::ICPUImage::ESCF_1_BIT,
+					.format = asset::EF_R16G16B16A16_SFLOAT,
+					.extent = {WIN_W, WIN_H, 1},
+					.mipLevels = 1,
+					.arrayLayers = 1,
+					.flags = IImage::ECF_NONE,
+					.usage = core::bitflag(asset::IImage::EUF_STORAGE_BIT) | asset::IImage::EUF_TRANSFER_SRC_BIT
+				}
+			});
+		if (!dummyImg || !m_device->allocate(dummyImg->getMemoryReqs(), dummyImg.get()).isValid())
+			return logFail("Could not create HDR Image");
```

Comment:
you don't need to create a dummy image to write into descriptor set, write the swapchain image instead